### PR TITLE
Sharing block header data around in a Portal history network (PoC)

### PR DESCRIPTION
**Pull request by kdeme**
- Rework PortalStream to have an instance per PortalProtocol (this
needs to be improved eventually). Each instance uses the same
UtpDiscv5Protocol instance.
- Add processContent on receival of accepted data
- Add dumb neighborhoodGossip: dumb in the sense that it only
offers one piece of content at a time.
- Add to / adjust populate_db to also allow for propagation of
the data and add debug rpc call: portal_history_propagate
- Add eth_rpc_client
- Add eth_getBlockbyHash (no txs or uncles) to eth API
- Add additional test to test_portal_testnet which loads 5 block
headers to 1 node, and offers this data to few nodes, which should
propagate it over the network further. Next query every node for
this data.

I'm not too pleased with the  current `PortalStream` vs `PortalProtocol` architecture but it will do for now. It will also need some additional checks still on the stream.transport in case it isn't used.

The eth rpc is also a bit of a mess, because I'm reusing code of nimbus, which is a bit of a clutter and imports all kinds of stuff that we don't need.

Further, there is some json-rpc error (RpcPostError) which I'm hitting when I'm doing too many calls. Need to further investigate but doesn't block this PR.

And then as last, Radius of all nodes is still set to max. I've played around with smaller values locally, and it still works. However, I'd have to think on how to incorporate this. Perhaps with some hidden cli argument for now (the eventual cli option would not be Radius but rather storage size or so, of which we can then derive an appropriate radius. But that only works on a real network with all the data)

### CODE REVIEW

**Review by KonradStaniec — APPROVED** [review]
I agree with sentiment

> I'm not too pleased with the current PortalStream vs PortalProtocol architecture

As it is sometimes hard to follow which callback leads to which part of logic, but we can improve in separate pr-s.

In general in nice to have whole flow implemented and tested 👍

File: fluffy/network/wire/portal_stream.nim [hunk 1]
```diff
@@ -121,6 +133,27 @@ proc readAndClose(socket: UtpSocket[NodeAddress], stream: PortalStream) {.async.
...
     socket.close()
 
+proc new*(
+    T: type PortalStream,
+    contentHandler: ContentHandlerCallback,
+    udata: ref,
+    connectionTimeout = defaultConnectionTimeout,
+    readTimeout = defaultReadTimeout,
+    rng = newRng()): T =
+  GC_ref(udata)
```
[the remark this answers is not in the document]
Comment (kdeme) [reply fluffy/network/wire/portal_stream.nim:143]: It's there because udata gets cast to a pointer which is stored in the `PortalStream` and that udata will get cleaned up by the GC when there are no more references to it (the pointer doesn't count, it is an untraced reference). So for the pointer a manual GC_ref is done to increase the ref count.

In practice, here it is not really an issue as the udata that is passed, the `PortalProtocol`, never goes out of scope until the program ends. We also don't have a location where to do the `GC_unref`, but in chronos the same thing is done for datagram and stream iirc.

File: fluffy/fluffy.nim [hunk 2]
```diff
@@ -75,16 +75,31 @@ proc run(config: PortalConf) {.raises: [CatchableError, Defect].} =
...
 
+  # One instance of UtpDiscv5Protocol is shared over all the PortalStreams.
+  let
+    socketConfig = SocketConfig.init(
+      incomingSocketReceiveTimeout = none(Duration))
+    streamTransport = UtpDiscv5Protocol.new(
+      d,
+      utpProtocolId,
+      registerIncomingSocketCallback(@[
+        stateNetwork.portalProtocol.stream,
```
[the remark this answers is not in the document]
Comment (kdeme) [reply fluffy/fluffy.nim:93]: Yeah, pretty much... :/
